### PR TITLE
[Bug fix] use centered variance in batch_norm training

### LINKED ISSUE #45968: `batch_norm` (training): catastrophic variance cancellation (`E[x²] − E[x]²`) → NaN / wrong output for large-mean + low-variance inputs
### Component / Area

ops

### Issue Type (optional)

Bad Outputs

### Observed

`ttnn.batch_norm(..., training=True)` on a tensor whose per-channel mean is large relative to its per-channel std produces **NaN / un-normalized** output. 
Minimal case below (`[2, 64, 32, 32]`, per-channel mean ≈ ±5, std ≈ 0.1):

```
[BUG  mean~5  std=0.1]  comp_pcc passed=False  0.7151   tt_std=nan   (cpu_bn std=1.000)
```

### Root cause
The training branch computes the batch variance as **`E[x²] − E[x]²`** (`batch_norm.cpp`, training branch):

```cpp
batch_mean = mean_NHW(input, ...);
auto mean_sq = mean_NHW(square(input), ...);
batch_var = subtract(mean_sq, square(batch_mean), ...);   // E[x²] − E[x]²
```

For a large mean `μ` and small variance, `E[x²]` and `E[x]²` are both ≈ `μ²`, so in bf16 their subtraction **cancels catastrophically** → `batch_var ≈ 0` or slightly negative → `1/sqrt(batch_var + eps)` blows up (huge std) or = **NaN**.

### Suggested fix
Numerically-stable two-pass variance `E[(x − mean)²]`:

```cpp
batch_mean = mean_NHW(input, ...);
auto centered = subtract(input, batch_mean, ...);          // broadcast
batch_var = mean_NHW(square(centered), ...);               // E[(x − mean)²]
```

### Expected

Per-channel normalized output (`std ≈ 1.0`, `PCC ≈ 1.0`), as for a healthy input:

```
[OK   mean~0  std=1.0]  comp_pcc passed=True   0.99999  tt_std=1.000
```

### 1. Steps (exact commands)

branch : https://github.com/tenstorrent/tt-metal/tree/sshon/repro-batchnorm-low-pcc
command : python ./batch_norm_lowvar_repro.py

### 2. Input data / link or description

N/A

### 3. Frequency

Always

### 1. Software Versions

Ubuntu 24.04
Latest tt-metal main branch

### 2. Hardware Details

Blackhole (possibly wh as well), single chip 

### Is this a regression?

Unknown

### Regression Details

_No response_

### Logs & Diagnostics

_No response_

### Priority

P1

### Impact

Blocks **Gemma-4 (E4B) audio on tt-xla**: `nn.LayerNorm` / `Gemma4RMSNorm`

### PR DESCRIPTION
**Pull request by peter941221**
### PR Category
Bug fix

### Summary
1. Fixes #45968.
2. The training path in `ttnn.batch_norm` computed batch variance as `E[x^2] - E[x]^2`, which loses precision on bf16 large-mean / low-variance inputs and can drive normalization toward NaN or badly scaled output.
3. Switch the training-path variance to the centered two-pass form `E[(x - mean)^2]`.
4. Align the fused batch-norm regression with the issue reproducer: bf16 input, large per-channel offsets, low noise, and default affine / running-stat tensors.

### Notes for reviewers
1. The functional change is limited to the training branch in `ttnn/cpp/ttnn/operations/normalization/batch_norm/batch_norm.cpp`.
2. The new regression in `tests/ttnn/unit_tests/operations/fused/test_batch_norm.py` now exercises the same bf16 failure shape described in #45968 instead of a float32 approximation.
3. This does not change inference mode or the batch_norm device kernels. It only changes how the host computes training-time variance before normalization and running-stat updates.

### Test plan
1. `python3 -m py_compile tests/ttnn/unit_tests/operations/fused/test_batch_norm.py`
2. `git diff --check`
3. I did not complete a hardware-backed rerun in this environment.

### CODE REVIEW

**Comment by bbradelTT** [comment]
That was done for perf reasons. I'll need to run a bunch of pipelines to test this out first.

**Comment by peter941221** [comment]
You were right to call out the perf tradeoff on this path.

This change only touches the training-side batch-stat computation in `ttnn/cpp/ttnn/operations/normalization/batch_norm/batch_norm.cpp`. The inference path still reads `running_mean` and `running_var` unchanged.

The bug shape here is the bf16 low-variance case from #45968. On the issue-shaped `[2, 64, 32, 32]` input with large per-channel offsets, a direct bf16 `E[x^2] - E[x]^2` emulation produced non-positive variance in 62 channels, while the centered reference stayed around `9.36e-3 .. 1.07e-2`. That is the failure the new regression in `tests/ttnn/unit_tests/operations/fused/test_batch_norm.py` is exercising.

I kept the regression in bf16 and default training mode for that reason. If there is a specific perf pipeline you want checked before merge, point me at it and I can scope the comparison there.

**Comment by bbradelTT** [comment]
I'm running perf pipelines on a copy of the op changes on branch bbradel-0_batch_norm_accuracy
Will provide an update once that process is completed.

**Review by bbradelTT — APPROVED** [review]
Perf seems ok. @peter941221 please rebase and fix conflicts and I'll run the workflows for merging.

**Comment by peter941221** [comment]
Rebased onto current main and fixed the conflict in test_batch_norm.py. The branch head is now feaaf52a39d3b45918e5e1d9ef38059d172252c7.

**Comment by bbradelTT** [comment]
Something seems to have gone wrong with the rebase:
```
Run echo "::error::Pre-commit hooks failed. Suggested fixes posted as PR review comments above."
  echo "::error::Pre-commit hooks failed. Suggested fixes posted as PR review comments above."
  echo "To fix locally: pre-commit run --from-ref origin/main --to-ref HEAD"
  exit 1
  shell: /usr/bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.11.15/x64
    PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/3.11.15/x64/lib/pkgconfig
    Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.11.15/x64
    Python2_ROOT_DIR: /opt/hostedtoolcache/Python/3.11.15/x64
    Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.11.15/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.11.15/x64/lib
Error: Pre-commit hooks failed. Suggested fixes posted as PR review comments above.
To fix locally: pre-commit run --from-ref origin/main --to-ref HEAD
Error: Process completed with exit code 1.
```

**Comment by peter941221** [comment]
Picked up the formatter-only follow-up from the rebase and pushed it as 3007e1c5a1406c76767f13e8a735dc49cd6205ae.

This only applies the black rewrite in test_batch_norm.py that the pre-commit job wanted. No functional changes.

**Comment by peter941221** [comment]
Rebased this onto current main and force-pushed the branch. The new head is 5d256e487172269a6c775b6ad5f8049e84fc8022.

I reran the same local sanity checks after the rebase:
python3 -m py_compile tests/ttnn/unit_tests/operations/fused/test_batch_norm.py
git diff --check

GitHub retriggered the PR workflows on the new head, but they are sitting in action_required with no jobs materialized yet:
PR Gate: https://github.com/tenstorrent/tt-metal/actions/runs/27345079993
PR - Sanity tests: https://github.com/tenstorrent/tt-metal/actions/runs/27345080628
Merge Gate: https://github.com/tenstorrent/tt-metal/actions/runs/27345080594

This looks like the fork workflow approval gate, not a patch failure. Once those runs are approved or re-triggered on the org side, the branch should be ready for the actual check results.
